# Thesis Codes  by Korkut Anapa

#  CLEANING CYCLE DESIGNER BY GOAL PROGRAMMING

## Optimization Model
Create a software that can design a cycle according to customer preferences

In [1]:
import numpy as np
import pandas as pd
from matplotlib import pyplot
from numpy import mean
from numpy import std
from sklearn.compose import make_column_selector as selector
from sklearn.datasets import make_regression
from sklearn.metrics import mean_squared_error
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import RepeatedKFold
from sklearn.model_selection import train_test_split
from xgboost import XGBRegressor
from sklearn.linear_model import LinearRegression

In [2]:
XX_all= ['10KUSum of BEKLEME',
'10KUSum of FAN',
'10KUSum of FAN / KLAPE',
'10KUSum of GAUDI_offset',
'10KUSum of KAPI ACMA',
'10KUSum of SURE',
'1OYMax of DEVIR',
'1OYMax of TEMP',
'1OYSum of ALT_C',
'1OYSum of BEKLEME',
'1OYSum of CIRC_TIME',
'1OYSum of kapalı',
'1OYSum of SURE',
'1OYSum of TEPE_C',
'1OYSum of WI',
'2AYMax of DEVIR',
'2AYMax of TEMP',
'2AYSum of ALT_C',
'2AYSum of kapalı',
'2AYSum of SURE',
'2AYSum of TEPE_C',
'2AYSum of WI',
'4SGDMax of DEVIR',
'4SGDSum of ALT_C',
'4SGDSum of SURE',
'4SGDSum of TEPE_C',
'4SGDSum of ÜST_C',
'4SGDSum of WI',
'7SCD2Max of DEVIR',
'7SCD2Max of TEMP',
'7SCD2Sum of ALT_C',
'7SCD2Sum of kapalı',
'7SCD2Sum of SURE',
'8DSBSum of kapalı']


In [3]:
XX_CPI=['1OYSum of SURE', '2AYSum of kapalı', '2AYSum of WI', '2AYMax of DEVIR',
       '2AYMax of TEMP', '4SGDSum of WI', '4SGDMax of DEVIR',
       '4SGDSum of ÜST_C', '4SGDSum of TEPE_C', '7SCD2Sum of kapalı',
       '7SCD2Max of DEVIR', '7SCD2Max of TEMP', '7SCD2Sum of ALT_C',
       '10KUSum of SURE', '10KUSum of FAN', '10KUSum of KAPI ACMA',
       '10KUSum of FAN / KLAPE', '10KUSum of GAUDI_offset']

In [4]:
XX_DPI=['1OYSum of CIRC_TIME', '1OYSum of WI', '1OYSum of TEPE_C',
       '2AYSum of kapalı', '2AYSum of SURE', '2AYSum of WI', '2AYMax of DEVIR',
       '2AYMax of TEMP', '2AYSum of TEPE_C', '4SGDSum of WI',
       '4SGDSum of ALT_C', '4SGDSum of ÜST_C', '7SCD2Sum of kapalı',
       '7SCD2Max of DEVIR', '7SCD2Max of TEMP', '7SCD2Sum of ALT_C',
       '10KUSum of SURE', '10KUSum of BEKLEME', '10KUSum of KAPI ACMA']

In [5]:
XX_EC=['1OYSum of kapalı', '1OYSum of SURE', '1OYSum of CIRC_TIME',
       '1OYSum of WI', '1OYSum of ALT_C', '2AYSum of kapalı', '2AYSum of SURE',
       '2AYMax of TEMP', '2AYSum of ALT_C', '2AYSum of TEPE_C',
       '4SGDSum of SURE', '4SGDSum of WI', '4SGDMax of DEVIR',
       '4SGDSum of ÜST_C', '4SGDSum of TEPE_C', '7SCD2Sum of kapalı',
       '7SCD2Sum of SURE', '7SCD2Max of TEMP', '7SCD2Sum of ALT_C',
       '10KUSum of BEKLEME', '10KUSum of FAN', '10KUSum of KAPI ACMA']

In [6]:
XX_WC=['1OYSum of kapalı', '1OYSum of WI', '1OYMax of DEVIR', '1OYMax of TEMP',
       '2AYSum of kapalı', '2AYSum of WI', '2AYMax of DEVIR',
       '4SGDSum of SURE', '4SGDSum of WI', '4SGDSum of TEPE_C',
       '7SCD2Sum of kapalı', '7SCD2Sum of SURE', '7SCD2Max of DEVIR',
       '7SCD2Sum of ALT_C', '8DSBSum of kapalı', '10KUSum of SURE',
       '10KUSum of GAUDI_offset']

In [7]:
XX_T=['1OYSum of SURE', '1OYSum of BEKLEME', '2AYSum of kapalı',
       '2AYSum of SURE', '2AYSum of ALT_C', '4SGDSum of WI',
       '7SCD2Sum of SURE', '7SCD2Max of DEVIR', '8DSBSum of kapalı',
       '10KUSum of SURE']

In [8]:
df_ref = pd.read_excel('D:/THESIS_DATA/ANALYSIS_DATA_THESIS/DB_SO_60.xlsx')

In [9]:
data=['CPI','DPI','EC','T','WC','1OYSum of kapalı' , '1OYSum of SURE' , '1OYSum of CIRC_TIME' , '1OYSum of WI' , '1OYMax of DEVIR',
      '1OYMax of TEMP' , '1OYSum of ALT_C' , '1OYSum of ÜST_C' , '1OYSum of TEPE_C' , '1OYSum of ZONE_C' , '1OYSum of BEKLEME' ,
      '1OYSum of FAN' , '1OYSum of KAPI ACMA' , '1OYSum of FAN / KLAPE' , '1OYSum of GAUDI_offset' ,
      '1OYSum of GAUDI_nonoffset' , '2AYSum of kapalı' , '2AYSum of SURE' , '2AYSum of CIRC_TIME' , '2AYSum of WI' ,
      '2AYMax of DEVIR' , '2AYMax of TEMP' , '2AYSum of ALT_C' , '2AYSum of ÜST_C' , '2AYSum of TEPE_C' , '2AYSum of ZONE_C' , 
      '2AYSum of BEKLEME' , '2AYSum of FAN' , '2AYSum of KAPI ACMA' , '2AYSum of FAN / KLAPE' , '2AYSum of GAUDI_offset' , 
      '2AYSum of GAUDI_nonoffset' , '3MFTSum of kapalı' , '3MFTSum of SURE' , '3MFTSum of CIRC_TIME' , '3MFTSum of WI' ,
      '3MFTMax of DEVIR' , '3MFTMax of TEMP' , '3MFTSum of ALT_C' , '3MFTSum of ÜST_C' , '3MFTSum of TEPE_C' , 
      '3MFTSum of ZONE_C' , '3MFTSum of BEKLEME' , '3MFTSum of FAN', '3MFTSum of KAPI ACMA' , '3MFTSum of FAN / KLAPE' , 
      '3MFTSum of GAUDI_offset' , '3MFTSum of GAUDI_nonoffset' , '4SGDSum of kapalı' , '4SGDSum of SURE' , 
      '4SGDSum of CIRC_TIME', '4SGDSum of WI' , '4SGDMax of DEVIR' , '4SGDMax of TEMP' , '4SGDSum of ALT_C' , 
      '4SGDSum of ÜST_C' , '4SGDSum of TEPE_C' , '4SGDSum of ZONE_C' , '4SGDSum of BEKLEME' , '4SGDSum of FAN' , 
      '4SGDSum of KAPI ACMA' , '4SGDSum of FAN / KLAPE' , '4SGDSum of GAUDI_offset' , '4SGDSum of GAUDI_nonoffset' , 
      '5ERSum of kapalı' , '5ERSum of SURE' , '5ERSum of CIRC_TIME' , '5ERSum of WI' , '5ERMax of DEVIR' , '5ERMax of TEMP' ,
      '5ERSum of ALT_C' , '5ERSum of ÜST_C' , '5ERSum of TEPE_C' , '5ERSum of ZONE_C' , '5ERSum of BEKLEME' ,
      '5ERSum of FAN' , '5ERSum of KAPI ACMA' , '5ERSum of FAN / KLAPE' , '5ERSum of GAUDI_offset' , 
      '5ERSum of GAUDI_nonoffset' , '6SCD1Sum of kapalı' , '6SCD1Sum of SURE' , '6SCD1Sum of CIRC_TIME' , '6SCD1Sum of WI' , 
      '6SCD1Max of DEVIR' , '6SCD1Max of TEMP' , '6SCD1Sum of ALT_C' , '6SCD1Sum of ÜST_C' , '6SCD1Sum of TEPE_C' , 
      '6SCD1Sum of ZONE_C' , '6SCD1Sum of BEKLEME' , '6SCD1Sum of FAN' , '6SCD1Sum of KAPI ACMA' , '6SCD1Sum of FAN / KLAPE' ,
      '6SCD1Sum of GAUDI_offset' , '6SCD1Sum of GAUDI_nonoffset' , '7SCD2Sum of kapalı' , '7SCD2Sum of SURE' , 
      '7SCD2Sum of CIRC_TIME' , '7SCD2Sum of WI' , '7SCD2Max of DEVIR' , '7SCD2Max of TEMP' , '7SCD2Sum of ALT_C' , 
      '7SCD2Sum of ÜST_C' , '7SCD2Sum of TEPE_C' , '7SCD2Sum of ZONE_C' , '7SCD2Sum of BEKLEME' , '7SCD2Sum of FAN' , 
      '7SCD2Sum of KAPI ACMA' , '7SCD2Sum of FAN / KLAPE' , '7SCD2Sum of GAUDI_offset' , '7SCD2Sum of GAUDI_nonoffset' , 
      '8DSBSum of kapalı' , '8DSBSum of SURE' , '8DSBSum of CIRC_TIME' , '8DSBSum of WI' , '8DSBMax of DEVIR' , 
      '8DSBMax of TEMP' , '8DSBSum of ALT_C' , '8DSBSum of ÜST_C', '8DSBSum of TEPE_C' , '8DSBSum of ZONE_C' ,
      '8DSBSum of BEKLEME' , '8DSBSum of FAN' , '8DSBSum of KAPI ACMA' , '8DSBSum of FAN / KLAPE' , '8DSBSum of GAUDI_offset' ,
      '8DSBSum of GAUDI_nonoffset', '9RYSum of kapalı' , '9RYSum of SURE' , '9RYSum of CIRC_TIME' , '9RYSum of WI' , 
      '9RYMax of DEVIR' , '9RYMax of TEMP' , '9RYSum of ALT_C' , '9RYSum of ÜST_C', '9RYSum of TEPE_C' , '9RYSum of ZONE_C' ,
      '9RYSum of BEKLEME' , '9RYSum of FAN' , '9RYSum of KAPI ACMA' , '9RYSum of FAN / KLAPE' , '9RYSum of GAUDI_offset' ,
      '9RYSum of GAUDI_nonoffset' , '10KUSum of kapalı' , '10KUSum of SURE' , '10KUSum of CIRC_TIME' , '10KUSum of WI' , 
      '10KUMax of DEVIR' , '10KUMax of TEMP' , '10KUSum of ALT_C' , '10KUSum of ÜST_C', '10KUSum of TEPE_C' ,
      '10KUSum of ZONE_C' , '10KUSum of BEKLEME' , '10KUSum of FAN' , '10KUSum of KAPI ACMA' , '10KUSum of FAN / KLAPE' ,
      '10KUSum of GAUDI_offset', '10KUSum of GAUDI_nonoffset']

In [10]:
indvar=['1OYSum of kapalı' , '1OYSum of SURE' , '1OYSum of CIRC_TIME' , '1OYSum of WI' , '1OYMax of DEVIR',
      '1OYMax of TEMP' , '1OYSum of ALT_C' , '1OYSum of ÜST_C' , '1OYSum of TEPE_C' , '1OYSum of ZONE_C' , '1OYSum of BEKLEME' ,
      '1OYSum of FAN' , '1OYSum of KAPI ACMA' , '1OYSum of FAN / KLAPE' , '1OYSum of GAUDI_offset' ,
      '1OYSum of GAUDI_nonoffset' , '2AYSum of kapalı' , '2AYSum of SURE' , '2AYSum of CIRC_TIME' , '2AYSum of WI' ,
      '2AYMax of DEVIR' , '2AYMax of TEMP' , '2AYSum of ALT_C' , '2AYSum of ÜST_C' , '2AYSum of TEPE_C' , '2AYSum of ZONE_C' , 
      '2AYSum of BEKLEME' , '2AYSum of FAN' , '2AYSum of KAPI ACMA' , '2AYSum of FAN / KLAPE' , '2AYSum of GAUDI_offset' , 
      '2AYSum of GAUDI_nonoffset' , '3MFTSum of kapalı' , '3MFTSum of SURE' , '3MFTSum of CIRC_TIME' , '3MFTSum of WI' ,
      '3MFTMax of DEVIR' , '3MFTMax of TEMP' , '3MFTSum of ALT_C' , '3MFTSum of ÜST_C' , '3MFTSum of TEPE_C' , 
      '3MFTSum of ZONE_C' , '3MFTSum of BEKLEME' , '3MFTSum of FAN', '3MFTSum of KAPI ACMA' , '3MFTSum of FAN / KLAPE' , 
      '3MFTSum of GAUDI_offset' , '3MFTSum of GAUDI_nonoffset' , '4SGDSum of kapalı' , '4SGDSum of SURE' , 
      '4SGDSum of CIRC_TIME', '4SGDSum of WI' , '4SGDMax of DEVIR' , '4SGDMax of TEMP' , '4SGDSum of ALT_C' , 
      '4SGDSum of ÜST_C' , '4SGDSum of TEPE_C' , '4SGDSum of ZONE_C' , '4SGDSum of BEKLEME' , '4SGDSum of FAN' , 
      '4SGDSum of KAPI ACMA' , '4SGDSum of FAN / KLAPE' , '4SGDSum of GAUDI_offset' , '4SGDSum of GAUDI_nonoffset' , 
      '5ERSum of kapalı' , '5ERSum of SURE' , '5ERSum of CIRC_TIME' , '5ERSum of WI' , '5ERMax of DEVIR' , '5ERMax of TEMP' ,
      '5ERSum of ALT_C' , '5ERSum of ÜST_C' , '5ERSum of TEPE_C' , '5ERSum of ZONE_C' , '5ERSum of BEKLEME' ,
      '5ERSum of FAN' , '5ERSum of KAPI ACMA' , '5ERSum of FAN / KLAPE' , '5ERSum of GAUDI_offset' , 
      '5ERSum of GAUDI_nonoffset' , '6SCD1Sum of kapalı' , '6SCD1Sum of SURE' , '6SCD1Sum of CIRC_TIME' , '6SCD1Sum of WI' , 
      '6SCD1Max of DEVIR' , '6SCD1Max of TEMP' , '6SCD1Sum of ALT_C' , '6SCD1Sum of ÜST_C' , '6SCD1Sum of TEPE_C' , 
      '6SCD1Sum of ZONE_C' , '6SCD1Sum of BEKLEME' , '6SCD1Sum of FAN' , '6SCD1Sum of KAPI ACMA' , '6SCD1Sum of FAN / KLAPE' ,
      '6SCD1Sum of GAUDI_offset' , '6SCD1Sum of GAUDI_nonoffset' , '7SCD2Sum of kapalı' , '7SCD2Sum of SURE' , 
      '7SCD2Sum of CIRC_TIME' , '7SCD2Sum of WI' , '7SCD2Max of DEVIR' , '7SCD2Max of TEMP' , '7SCD2Sum of ALT_C' , 
      '7SCD2Sum of ÜST_C' , '7SCD2Sum of TEPE_C' , '7SCD2Sum of ZONE_C' , '7SCD2Sum of BEKLEME' , '7SCD2Sum of FAN' , 
      '7SCD2Sum of KAPI ACMA' , '7SCD2Sum of FAN / KLAPE' , '7SCD2Sum of GAUDI_offset' , '7SCD2Sum of GAUDI_nonoffset' , 
      '8DSBSum of kapalı' , '8DSBSum of SURE' , '8DSBSum of CIRC_TIME' , '8DSBSum of WI' , '8DSBMax of DEVIR' , 
      '8DSBMax of TEMP' , '8DSBSum of ALT_C' , '8DSBSum of ÜST_C', '8DSBSum of TEPE_C' , '8DSBSum of ZONE_C' ,
      '8DSBSum of BEKLEME' , '8DSBSum of FAN' , '8DSBSum of KAPI ACMA' , '8DSBSum of FAN / KLAPE' , '8DSBSum of GAUDI_offset' ,
      '8DSBSum of GAUDI_nonoffset', '9RYSum of kapalı' , '9RYSum of SURE' , '9RYSum of CIRC_TIME' , '9RYSum of WI' , 
      '9RYMax of DEVIR' , '9RYMax of TEMP' , '9RYSum of ALT_C' , '9RYSum of ÜST_C', '9RYSum of TEPE_C' , '9RYSum of ZONE_C' ,
      '9RYSum of BEKLEME' , '9RYSum of FAN' , '9RYSum of KAPI ACMA' , '9RYSum of FAN / KLAPE' , '9RYSum of GAUDI_offset' ,
      '9RYSum of GAUDI_nonoffset' , '10KUSum of kapalı' , '10KUSum of SURE' , '10KUSum of CIRC_TIME' , '10KUSum of WI' , 
      '10KUMax of DEVIR' , '10KUMax of TEMP' , '10KUSum of ALT_C' , '10KUSum of ÜST_C', '10KUSum of TEPE_C' ,
      '10KUSum of ZONE_C' , '10KUSum of BEKLEME' , '10KUSum of FAN' , '10KUSum of KAPI ACMA' , '10KUSum of FAN / KLAPE' ,
      '10KUSum of GAUDI_offset', '10KUSum of GAUDI_nonoffset']

In [11]:
df=df_ref[data]

In [12]:
df_ver_0=df.drop_duplicates(keep='last',subset=indvar)

In [13]:
df_ver_0

CPI        DPI        EC           T         WC  1OYSum of kapalı  \
3    3.220000  76.078000  0.500000  265.000000  10.000000                 2   
4    3.280000  78.235000  0.590000  265.000000   9.200000                 3   
5    3.270000  81.176000  0.570000  275.000000   9.480000                 8   
6    3.300000  77.647000  0.580000  250.000000   9.410000                 8   
8    3.270000  82.824000  0.650000  250.000000   9.770000                 8   
..        ...        ...       ...         ...        ...               ...   
149  3.300662  84.105960  1.046000  226.000000  12.113333                 0   
150  3.358667  87.000000  0.951622  220.000000   9.500000                 0   
151  3.550000  80.000000  0.864750  210.000000  12.000000                 0   
152  3.230864  86.316872  0.861667  227.666667  10.726667                 0   
153  3.224277  85.260116  0.906000  252.500000  11.285000                 0   

     1OYSum of SURE  1OYSum of CIRC_TIME  1OYSum of WI  1OYMax of DEVIR  ...  \
3                48                   43           4.8             2800  ...   
4                49                   44           4.8             2800  ...   
5                54                   49           4.8             2800  ...   
6                54                   49           4.8             2800  ...   
8                55                   49           7.1             2800  ...   
..              ...                  ...           ...              ...  ...   
149              23                   20           4.6             2800  ...   
150              19                   16           4.0             2800  ...   
151               8                    6           3.5             2800  ...   
152              19                   16           4.0             2800  ...   
153              24                   21           4.0             2800  ...   

     10KUSum of ALT_C  10KUSum of ÜST_C  10KUSum of TEPE_C  10KUSum of ZONE_C  \
3                   0                 0                  0                  0   
4                   0                 0                  0                  0   
5                   0                 0                  0                  0   
6                   0                 0                  0                  0   
8                   0                 0                  0                  0   
..                ...               ...                ...                ...   
149                 0                 0                  0                  0   
150                 0                 0                  0                  0   
151                 0                 0                  0                  0   
152                 0                 0                  0                  0   
153                 0                 0                  0                  0   

     10KUSum of BEKLEME  10KUSum of FAN  10KUSum of KAPI ACMA  \
3                    50              57                     1   
4                    30              63                     1   
5                    30              80                     1   
6                    30              63                     1   
8                    30              63                     1   
..                  ...             ...                   ...   
149                 102               0                     0   
150                  20              73                     0   
151                 123               0                     0   
152                  10              98                     0   
153                  10              98                     0   

     10KUSum of FAN / KLAPE  10KUSum of GAUDI_offset  \
3                        30                        0   
4                        29                        0   
5                        38                        0   
6                        29                        0   
8                        29                        0   
..        

In [14]:
y_CPI = df_ver_0['CPI']
y_DPI = df_ver_0['DPI']
y_EC=df_ver_0['EC']
y_T=df_ver_0['T']
y_WC=df_ver_0['WC']

In [15]:
X_CPI=df_ver_0[XX_CPI]
X_DPI=df_ver_0[XX_DPI]
X_EC=df_ver_0[XX_EC]
X_WC=df_ver_0[XX_WC]
X_T=df_ver_0[XX_T]
X_all=df_ver_0[XX_all]

In [16]:
y=y_CPI
X=X_CPI
model_CPI = LinearRegression()
model_CPI.fit(X, y)
cv = RepeatedKFold(n_splits=5, n_repeats=3, random_state=79)
scores1 = cross_val_score(model_CPI,X, y, scoring='neg_mean_absolute_error', cv=cv, n_jobs=-1, error_score='raise')
scores3 = cross_val_score(model_CPI,X, y, scoring='r2', cv=cv, n_jobs=-1, error_score='raise')
print(' %.2f %.0f' % (-mean(scores1), 100*mean(scores3)))

 0.08 64


In [17]:
y=y_DPI
X=X_DPI
model_DPI = LinearRegression()
model_DPI.fit(X, y)
cv = RepeatedKFold(n_splits=5, n_repeats=3, random_state=79)
scores1 = cross_val_score(model_DPI,X, y, scoring='neg_mean_absolute_error', cv=cv, n_jobs=-1, error_score='raise')
scores3 = cross_val_score(model_DPI,X, y, scoring='r2', cv=cv, n_jobs=-1, error_score='raise')
print(' %.2f %.0f' % (-mean(scores1), 100*mean(scores3)))

 4.01 60


In [18]:
y=y_EC
X=X_EC
model_EC = LinearRegression()
model_EC.fit(X, y)
cv = RepeatedKFold(n_splits=5, n_repeats=3, random_state=79)
scores1 = cross_val_score(model_EC,X, y, scoring='neg_mean_absolute_error', cv=cv, n_jobs=-1, error_score='raise')
scores3 = cross_val_score(model_EC,X, y, scoring='r2', cv=cv, n_jobs=-1, error_score='raise')
print(' %.2f %.0f' % (-mean(scores1), 100*mean(scores3)))

 0.04 88


In [19]:
y=y_WC
X=X_WC
model_WC = LinearRegression()
model_WC.fit(X, y)
cv = RepeatedKFold(n_splits=5, n_repeats=3, random_state=79)
scores1 = cross_val_score(model_WC,X, y, scoring='neg_mean_absolute_error', cv=cv, n_jobs=-1, error_score='raise')
scores3 = cross_val_score(model_WC,X, y, scoring='r2', cv=cv, n_jobs=-1, error_score='raise')
print(' %.2f %.0f' % (-mean(scores1), 100*mean(scores3)))

 0.41 89


In [20]:
model_WC.coef_

array([ 3.54745381e-02,  8.80795329e-01,  3.68325835e-04, -1.49193183e-02,
       -3.54288192e-01,  9.99390234e-01,  5.99973440e-04,  3.30618874e-02,
        1.16513447e+00, -1.57243856e-01,  2.22478005e-01, -5.51177511e-02,
       -1.19688497e-03,  5.84132118e-02, -8.25626850e-01, -4.40253439e-03,
       -2.15232305e-03])

In [21]:
model_WC.intercept_

4.688052250462984

In [22]:
y=y_T
X=X_T
model_T = LinearRegression()
model_T.fit(X, y)
cv = RepeatedKFold(n_splits=5, n_repeats=3, random_state=79)
scores1 = cross_val_score(model_T,X, y, scoring='neg_mean_absolute_error', cv=cv, n_jobs=-1, error_score='raise')
scores3 = cross_val_score(model_T,X, y, scoring='r2', cv=cv, n_jobs=-1, error_score='raise')
print(' %.2f %.0f' % (-mean(scores1), 100*mean(scores3)))

 2.84 100


## SELECTED INDEPENDENT VARIABLES AND MODELS

In [23]:
df_base = pd.read_excel('D:/THESIS_DATA/ANALYSIS_DATA_THESIS/DB_SO_eco.xlsx')

In [24]:
X_base=df_base[XX_all]

In [25]:
X_base

10KUSum of BEKLEME  10KUSum of FAN  10KUSum of FAN / KLAPE  \
0                  80               0                       0   

   10KUSum of GAUDI_offset  10KUSum of KAPI ACMA  10KUSum of SURE  \
0                        0                     1               82   

   1OYMax of DEVIR  1OYMax of TEMP  1OYSum of ALT_C  1OYSum of BEKLEME  ...  \
0             2800              54             13.5                  0  ...   

   4SGDSum of SURE  4SGDSum of TEPE_C  4SGDSum of ÜST_C  4SGDSum of WI  \
0               14                  4                 6            2.6   

   7SCD2Max of DEVIR  7SCD2Max of TEMP  7SCD2Sum of ALT_C  7SCD2Sum of kapalı  \
0               2600                53                  6                   4   

   7SCD2Sum of SURE  8DSBSum of kapalı  
0                28                  0  

[1 rows x 34 columns]

In [26]:
X_test=df_base[XX_T]
X_test
model_T.predict(X_test)

array([235.73354432])

In [27]:
xl=[]
xu=[]

In [28]:
X=X_base
for term in X:
    xl.append(0.9*min(X_base[term]))
    xu.append(1.1*max(X_base[term]))

In [29]:
from pyomo.environ import *
from pyomo.opt import SolverFactory

In [30]:
#opt = SolverFactory('ipopt')

In [31]:
opt = SolverFactory('cplex_direct')
# Create an object of a concrete model
model = ConcreteModel()
# Define the decision variables
model.x1 = Var(within = NonNegativeReals)
model.x2 = Var(within = NonNegativeReals)
model.x3 = Var(within = NonNegativeReals)
model.x4 = Var(within = NonNegativeReals)
model.x5 = Var(within = NonNegativeReals)
model.x6 = Var(within = NonNegativeReals)
model.x7 = Var(within = NonNegativeReals)
model.x8 = Var(within = NonNegativeReals)
model.x9 = Var(within = NonNegativeReals)
model.x10 = Var(within = NonNegativeReals)
model.x11 = Var(within = NonNegativeReals)
model.x12 = Var(within = NonNegativeReals)
model.x13 = Var(within = NonNegativeReals)
model.x14 = Var(within = NonNegativeReals)
model.x15 = Var(within = NonNegativeReals)
model.x16 = Var(within = NonNegativeReals)
model.x17 = Var(within = NonNegativeReals)
model.x18 = Var(within = NonNegativeReals)
model.x19 = Var(within = NonNegativeReals)
model.x20 = Var(within = NonNegativeReals)
model.x21 = Var(within = NonNegativeReals)
model.x22 = Var(within = NonNegativeReals)
model.x23 = Var(within = NonNegativeReals)
model.x24 = Var(within = NonNegativeReals)
model.x25 = Var(within = NonNegativeReals)
model.x26 = Var(within = NonNegativeReals)
model.x27 = Var(within = NonNegativeReals)
model.x28 = Var(within = NonNegativeReals)
model.x29 = Var(within = NonNegativeReals)
model.x30 = Var(within = NonNegativeReals)
model.x31 = Var(within = NonNegativeReals)
model.x32 = Var(within = NonNegativeReals)
model.x33 = Var(within = NonNegativeReals)
model.x34 = Var(within = NonNegativeReals)

# Define the deviational variables
model.n1 = Var(within = NonNegativeReals)
model.p1 = Var(within = NonNegativeReals)
model.n2 = Var(within = NonNegativeReals)
model.p2 = Var(within = NonNegativeReals)
model.n3 = Var(within = NonNegativeReals)
model.p3 = Var(within = NonNegativeReals)
model.n4 = Var(within = NonNegativeReals)
model.p4 = Var(within = NonNegativeReals)
model.n5 = Var(within = NonNegativeReals)
model.p5 = Var(within = NonNegativeReals)

In [32]:
CPI0=model_CPI.intercept_
DPI0=model_DPI.intercept_
EC0=model_EC.intercept_
T0=model_T.intercept_
WC0=model_WC.intercept_

In [33]:
for i in range(1, 11):
    globals()[f"T{i}"] = model_T.coef_[i-1]

In [34]:
for i in range(1, 19):
    globals()[f"CPI{i}"] = model_CPI.coef_[i-1]

In [35]:
for i in range(1, 20):    
    globals()[f"DPI{i}"] = model_DPI.coef_[i-1]

In [36]:
for i in range(1, 23):    
    globals()[f"EC{i}"] = model_EC.coef_[i-1]

In [37]:
for i in range(1, 18):    
    globals()[f"WC{i}"] = model_WC.coef_[i-1]

In [38]:
CPI=3.3
DPI=80
EC=0.6
T=360
WC=9.5

In [39]:
model.l=Var(within = NonNegativeReals)

In [40]:
# Define the objective function
model.obj = Objective(expr = model.l)

# Define the constraints
model.con1= Constraint(expr = (2 / CPI) * model.n1 <= model.l)
model.con2= Constraint(expr = (2 / DPI) * model.n2 <= model.l)
model.con3= Constraint(expr = (3 / EC) * model.p3 <= model.l)
model.con4= Constraint(expr = (1 / T) * model.p4 <= model.l)
model.con5= Constraint(expr = (1 / WC) * model.p5 <= model.l)

model.con6 = Constraint(expr = CPI0 + CPI1*model.x13+CPI2*model.x19+CPI3*model.x22+CPI4*model.x16+CPI5*model.x17+CPI6*model.x28+
                                CPI7*model.x23+CPI8*model.x27+CPI9*model.x26+CPI10*model.x32+CPI11*model.x29+CPI12*model.x30+
                        CPI13*model.x31+CPI14*model.x6+
                                CPI15*model.x2+CPI16*model.x5+CPI17*model.x3+CPI18*model.x4+model.n1 
                        -model.p1 == CPI)

model.con7 = Constraint(expr = DPI0 + DPI1*model.x11+DPI2*model.x15+DPI3*model.x14+DPI4*model.x19+DPI5*model.x20+DPI6*model.x22+
                                DPI7*model.x16+DPI8*model.x17+DPI9*model.x21+DPI10*model.x28+DPI11*model.x24+DPI12*model.x27+
                        DPI13*model.x32+DPI14*model.x29+
                                DPI15*model.x30+DPI16*model.x31+DPI17*model.x6+DPI18*model.x1+DPI19*model.x5+model.n2 
                        -model.p2 == DPI)

model.con8= Constraint(expr = EC0 + EC1*model.x12+EC2*model.x13+EC3*model.x11+EC4*model.x15+EC5*model.x9+EC6*model.x19+
                                EC7*model.x20+EC8*model.x17+EC9*model.x18+EC10*model.x21+EC11*model.x25+EC12*model.x28
                       +EC13*model.x23+EC14*model.x27+
                                EC15*model.x26+EC16*model.x32+EC17*model.x33+EC18*model.x30+EC19*model.x31+ EC20*model.x1 +
                       EC21*model.x2 + EC22*model.x5 + model.n3 -
                       model.p3 == EC)

model.con9= Constraint(expr = T0 + T1*model.x13+T2*model.x10+T3*model.x19+T4*model.x20+T5*model.x18+T6*model.x28+
                                T7*model.x33+T8*model.x29+T9*model.x34+T10*model.x6+model.n4 -model.p4 == T)

model.con10= Constraint(expr = WC0 + WC1*model.x12+WC2*model.x15+WC3*model.x7+WC4*model.x8+WC5*model.x19+WC6*model.x22+
                                WC7*model.x16+WC8*model.x25+WC9*model.x28+WC10*model.x26+WC11*model.x32+WC12*model.x33+
                                WC13*model.x29+WC14*model.x31+WC15*model.x34+WC16*model.x6+
                                WC17*model.x4+model.n5 -model.p5 == WC)




model.con11=Constraint(expr = model.x1>=xl[0])
model.con12=Constraint(expr = model.x1<=xu[0])
model.con13=Constraint(expr = model.x2>=xl[1])
model.con14=Constraint(expr = model.x2<=xu[1])
model.con15=Constraint(expr = model.x3>=xl[2])
model.con16=Constraint(expr = model.x3<=xu[2])
model.con17=Constraint(expr = model.x4>=xl[3])
model.con18=Constraint(expr = model.x4<=xu[3])
model.con19=Constraint(expr = model.x5>=xl[4])
model.con20=Constraint(expr = model.x5<=xu[4])
model.con21 =Constraint(expr = model.x6>=xl[5])
model.con22 =Constraint(expr = model.x6<=xu[5])
model.con23 =Constraint(expr = model.x7>=xl[6])
model.con24 =Constraint(expr = model.x7<=xu[6])
model.con25 =Constraint(expr = model.x8>=xl[7])
model.con26 =Constraint(expr = model.x8<=xu[7])
model.con27 =Constraint(expr = model.x9>=xl[8])
model.con28 =Constraint(expr = model.x9<=xu[8])
model.con29 =Constraint(expr = model.x10>=xl[9])
model.con30 =Constraint(expr = model.x10<=xu[9])
model.con31 =Constraint(expr = model.x11>=xl[10])
model.con32 =Constraint(expr = model.x11<=xu[10])
model.con33 =Constraint(expr = model.x12>=xl[11])
model.con34 =Constraint(expr = model.x12<=xu[11])
model.con35 =Constraint(expr = model.x13>=xl[12])
model.con36 =Constraint(expr = model.x13<=xu[12])
model.con37 =Constraint(expr = model.x14>=xl[13])
model.con38 =Constraint(expr = model.x14<=xu[13])
model.con39 =Constraint(expr = model.x15>=xl[14])
model.con40 =Constraint(expr = model.x15<=xu[14])
model.con41 =Constraint(expr = model.x16>=xl[15])
model.con42 =Constraint(expr = model.x16<=xu[15])
model.con43 =Constraint(expr = model.x17>=xl[16])
model.con44 =Constraint(expr = model.x17<=xu[16])
model.con45 =Constraint(expr = model.x18>=xl[17])
model.con46 =Constraint(expr = model.x18<=xu[17])
model.con47 =Constraint(expr = model.x19>=xl[18])
model.con48 =Constraint(expr = model.x19<=xu[18])
model.con49 =Constraint(expr = model.x20>=xl[19])
model.con50 =Constraint(expr = model.x20<=xu[19])
model.con51 =Constraint(expr = model.x21>=xl[20])
model.con52 =Constraint(expr = model.x21<=xu[20])
model.con53 =Constraint(expr = model.x22>=xl[21])
model.con54 =Constraint(expr = model.x22<=xu[21])
model.con55 =Constraint(expr = model.x23>=xl[22])
model.con56 =Constraint(expr = model.x23<=xu[22])
model.con57 =Constraint(expr = model.x24>=xl[23])
model.con58 =Constraint(expr = model.x24<=xu[23])
model.con59 =Constraint(expr = model.x25>=xl[24])
model.con60 =Constraint(expr = model.x25<=xu[24])
model.con61 =Constraint(expr = model.x26>=xl[25])
model.con62 =Constraint(expr = model.x26<=xu[25])
model.con63 =Constraint(expr = model.x27>=xl[26])
model.con64 =Constraint(expr = model.x27<=xu[26])
model.con65 =Constraint(expr = model.x28>=xl[27])
model.con66 =Constraint(expr = model.x28<=xu[27])
model.con67 =Constraint(expr = model.x29>=xl[28])
model.con68 =Constraint(expr = model.x29<=xu[28])
model.con69 =Constraint(expr = model.x30>=xl[29])
model.con70 =Constraint(expr = model.x30<=xu[29])
model.con71 =Constraint(expr = model.x31>=xl[30])
model.con72 =Constraint(expr = model.x31<=xu[30])
model.con73 =Constraint(expr = model.x32>=xl[31])
model.con74 =Constraint(expr = model.x32<=xu[31])
model.con75 =Constraint(expr = model.x33>=xl[32])
model.con76 =Constraint(expr = model.x33<=xu[32])
model.con77 =Constraint(expr = model.x34>=xl[33])
model.con78 =Constraint(expr = model.x34<=xu[33])



# Solve the Goal Programming problem
opt.solve(model)

{'Problem': [{'Name': '', 'Lower bound': 0.0, 'Upper bound': 0.0, 'Number of objectives': 1, 'Number of constraints': 78, 'Number of variables': 45, 'Number of binary variables': 0, 'Number of integer variables': 0, 'Number of continuous variables': 45, 'Number of nonzeros': None, 'Sense': 1}], 'Solver': [{'Name': 'CPLEX 22.1.0.0', 'Status': 'ok', 'Wallclock time': 0.0, 'Termination condition': 'optimal'}], 'Solution': [OrderedDict([('number of solutions', 0), ('number of solutions displayed', 0)])]}

In [41]:
# Print the values of the decision variables

print("CPI_TARGET = ", CPI)
print("DPI_TARGET = ", DPI)
print("EC_TARGET = ", EC)
print("T_TARGET = ", T)
print("WC_TARGET = ", WC)

print("x1  =",model.x1.value)
print("x2  =",model.x2.value)
print("x3  =",model.x3.value)
print("x4  =",model.x4.value)
print("x5  =",model.x5.value)
print("x6  =",model.x6.value)
print("x7  =",model.x7.value)
print("x8  =",model.x8.value)
print("x9  =",model.x9.value)
print("x10  =",model.x10.value)
print("x11  =",model.x11.value)
print("x12  =",model.x12.value)
print("x13  =",model.x13.value)
print("x14  =",model.x14.value)
print("x15  =",model.x15.value)
print("x16  =",model.x16.value)
print("x17  =",model.x17.value)
print("x18  =",model.x18.value)
print("x19  =",model.x19.value)
print("x19  =",model.x19.value)
print("x20  =",model.x20.value)
print("x21  =",model.x21.value)
print("x22  =",model.x22.value)
print("x23  =",model.x23.value)
print("x24  =",model.x24.value)
print("x25  =",model.x25.value)
print("x26  =",model.x26.value)
print("x27  =",model.x27.value)
print("x28  =",model.x28.value)
print("x29  =",model.x29.value)
print("x30  =",model.x30.value)
print("x31  =",model.x31.value)
print("x32  =",model.x32.value)
print("x33  =",model.x33.value)
print("x34  =",model.x34.value)

# Print the achieved values for each goal
if model.n1.value > 0:
    print("The first goal is underachieved by ",CPI-model.n1.value)
elif model.p1.value > 0:
    print("The first goal is overachieved by ", model.p1.value+CPI)
else:
    print("The first goal is fully satisfied",CPI)


if model.n2.value > 0:
    print("The second goal is underachieved by ",DPI-model.n2.value)
elif model.p2.value > 0:
    print("The second goal is overachieved by ",model.p2.value+DPI)
else:
    print("The second goal is fully satisfied",DPI)

if model.n3.value > 0:
    print("The third goal is underachieved by ",EC-model.n3.value)
elif model.p3.value > 0:
    print("The third goal is overachieved by ",model.p3.value+EC)
else:
    print("The third goal is fully satisfied",EC)
    
    
if model.n4.value > 0:
    print("The fourth goal is underachieved by ",T-model.n4.value)
elif model.p4.value > 0:
    print("The fourth goal is overachieved by ",model.p4.value+T)
else:
    print("The fourth goal is fully satisfied",T)
    
    
if model.n5.value > 0:
    print("The fifth goal is underachieved by ",WC-model.n5.value)
elif model.p5.value > 0:
    print("The fifth goal is overachieved by ",model.p5.value+WC)
else:
    print("The fifth goal is fully satisfied",WC)

CPI_TARGET =  3.3
DPI_TARGET =  80
EC_TARGET =  0.6
T_TARGET =  360
WC_TARGET =  9.5
x1  = 72.0
x2  = 0.0
x3  = 0.0
x4  = 0.0
x5  = 0.959705705744772
x6  = 73.8
x7  = 2520.0
x8  = 59.400000000000006
x9  = 14.850000000000001
x10  = 0.0
x11  = 36.0
x12  = 5.4
x13  = 47.300000000000004
x14  = 10.266666666666666
x15  = 3.6
x16  = 3300.0000000000005
x17  = 45.0
x18  = 25.069000000000003
x19  = 0.0
x19  = 0.0
x20  = 74.80000000000001
x21  = 18.261
x22  = 0.0
x23  = 2160.0
x24  = 1.8
x25  = 12.6
x26  = 3.6
x27  = 5.4
x28  = 2.3400000000000003
x29  = 2860.0000000000005
x30  = 47.7
x31  = 5.4
x32  = 3.6
x33  = 25.2
x34  = 0.0
The first goal is fully satisfied 3.3
The second goal is overachieved by  83.78391206160491
The third goal is underachieved by  0.5964167141389773
The fourth goal is underachieved by  235.52687236837107
The fifth goal is underachieved by  8.628667229865204


CPI_TARGET =  3.33
DPI_TARGET =  79
EC_TARGET =  0.7375
T_TARGET =  235
WC_TARGET =  9.5
x1  = 80.0
x2  = 0.0
x3  = 0.0
x4  = 0.0
x5  = 1.0
x6  = 82.0
x7  = 2800.0
x8  = 54.0
x9  = 13.5
x10  = 0.0
x11  = 40.0
x12  = 6.0
x13  = 43.0
x14  = 9.333333333333332
x15  = 4.0
x16  = 3000.0
x17  = 50.0
x18  = 22.79
x19  = 0.0
x19  = 0.0
x20  = 68.0
x21  = 20.29
x22  = 0.0
x23  = 2400.0
x24  = 2.0
x25  = 14.0
x26  = 4.0
x27  = 6.0
x28  = 2.6
x29  = 2600.0
x30  = 53.0
x31  = 6.0
x32  = 4.0
x33  = 28.0
x34  = 0.0
The first goal is underachieved by  3.329471703991409
The second goal is overachieved by  79.67219892321478
The third goal is overachieved by  0.737569427761937
The fourth goal is overachieved by  235.73354431511126
The fifth goal is overachieved by  9.537096402946908

CPI_TARGET =  3.15
DPI_TARGET =  75
EC_TARGET =  0.5
T_TARGET =  360
WC_TARGET =  9.5
x1  = 72.0
x2  = 0.0
x3  = 0.0
x4  = 0.0
x5  = 0.9
x6  = 73.8
x7  = 2520.0
x8  = 59.400000000000006
x9  = 14.850000000000001
x10  = 0.0
x11  = 36.0
x12  = 5.4
x13  = 47.300000000000004
x14  = 10.266666666666666
x15  = 3.6
x16  = 2952.139383693132
x17  = 45.0
x18  = 25.069000000000003
x19  = 0.0
x19  = 0.0
x20  = 74.80000000000001
x21  = 18.261
x22  = 0.0
x23  = 2160.0
x24  = 1.8
x25  = 14.211086732351431
x26  = 3.6
x27  = 5.4
x28  = 2.3400000000000003
x29  = 2860.0000000000005
x30  = 47.7
x31  = 5.4
x32  = 3.6
x33  = 25.2
x34  = 0.0
The first goal is fully satisfied 3.15
The second goal is overachieved by  78.86476867462797
The third goal is fully satisfied 0.5
The fourth goal is underachieved by  235.52687236837113
The fifth goal is underachieved by  8.473225667173734